// @n8n/.ai-workflow-builder-8gF1dZcr/dist/workflow-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowState = void 0;
const langgraph_1 = require("@langchain/langgraph");
function operationsReducer(current, update) {
    if (update === null) {
        return [];
    }
    if (!update || update.length === 0) {
        return current ?? [];
    }
    if (update.some((op) => op.type === 'clear')) {
        return update.filter((op) => op.type === 'clear').slice(-1);
    }
    if (!current && !update) {
        return [];
    }
    return [...(current ?? []), ...update];
}
exports.WorkflowState = langgraph_1.Annotation.Root({
    messages: (0, langgraph_1.Annotation)({
        reducer: langgraph_1.messagesStateReducer,
        default: () => [],
    }),
    workflowJSON: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => ({ nodes: [], connections: {} }),
    }),
    workflowOperations: (0, langgraph_1.Annotation)({
        reducer: operationsReducer,
        default: () => [],
    }),
    workflowContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
    }),
});
