// n8n-workflow/dist/cjs/errors/expression.error.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./abstract/execution-base.error"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ExpressionError = exports.EXPRESSION_DESCRIPTION_KEYS = exports.EXPRESSION_ERROR_TYPES = exports.EXPRESSION_ERROR_MESSAGES = void 0;
    const execution_base_error_1 = require("./abstract/execution-base.error");
    /**
     * Class for instantiating an expression error
     */
    // Expression error constants
    exports.EXPRESSION_ERROR_MESSAGES = {
        NODE_NOT_FOUND: 'Error finding the referenced node',
        NODE_REFERENCE_TEMPLATE: 'Make sure the node you referenced is spelled correctly and is a parent of this node',
        NO_EXECUTION_DATA: 'No execution data available',
    };
    exports.EXPRESSION_ERROR_TYPES = {
        PAIRED_ITEM_NO_CONNECTION: 'paired_item_no_connection',
    };
    exports.EXPRESSION_DESCRIPTION_KEYS = {
        NODE_NOT_FOUND: 'nodeNotFound',
        NO_NODE_EXECUTION_DATA: 'noNodeExecutionData',
        PAIRED_ITEM_NO_CONNECTION: 'pairedItemNoConnection',
        PAIRED_ITEM_NO_CONNECTION_CODE_NODE: 'pairedItemNoConnectionCodeNode',
    };
    class ExpressionError extends execution_base_error_1.ExecutionBaseError {
        constructor(message, options) {
            super(message, { cause: options?.cause, level: 'warning' });
            if (options?.description !== undefined) {
                this.description = options.description;
            }
            const allowedKeys = [
                'causeDetailed',
                'descriptionTemplate',
                'descriptionKey',
                'itemIndex',
                'messageTemplate',
                'nodeCause',
                'parameter',
                'runIndex',
                'type',
            ];
            if (options !== undefined) {
                if (options.functionality !== undefined) {
                    this.functionality = options.functionality;
                }
                Object.keys(options).forEach((key) => {
                    if (allowedKeys.includes(key)) {
                        this.context[key] = options[key];
                    }
                });
            }
        }
    }
    exports.ExpressionError = ExpressionError;
});
