// n8n-workflow/dist/esm/errors/expression.error.js
import { ExecutionBaseError } from './abstract/execution-base.error';
/**
 * Class for instantiating an expression error
 */
// Expression error constants
export const EXPRESSION_ERROR_MESSAGES = {
    NODE_NOT_FOUND: 'Error finding the referenced node',
    NODE_REFERENCE_TEMPLATE: 'Make sure the node you referenced is spelled correctly and is a parent of this node',
    NO_EXECUTION_DATA: 'No execution data available',
};
export const EXPRESSION_ERROR_TYPES = {
    PAIRED_ITEM_NO_CONNECTION: 'paired_item_no_connection',
};
export const EXPRESSION_DESCRIPTION_KEYS = {
    NODE_NOT_FOUND: 'nodeNotFound',
    NO_NODE_EXECUTION_DATA: 'noNodeExecutionData',
    PAIRED_ITEM_NO_CONNECTION: 'pairedItemNoConnection',
    PAIRED_ITEM_NO_CONNECTION_CODE_NODE: 'pairedItemNoConnectionCodeNode',
};
export class ExpressionError extends ExecutionBaseError {
    constructor(message, options) {
        super(message, { cause: options?.cause, level: 'warning' });
        if (options?.description !== undefined) {
            this.description = options.description;
        }
        const allowedKeys = [
            'causeDetailed',
            'descriptionTemplate',
            'descriptionKey',
            'itemIndex',
            'messageTemplate',
            'nodeCause',
            'parameter',
            'runIndex',
            'type',
        ];
        if (options !== undefined) {
            if (options.functionality !== undefined) {
                this.functionality = options.functionality;
            }
            Object.keys(options).forEach((key) => {
                if (allowedKeys.includes(key)) {
                    this.context[key] = options[key];
                }
            });
        }
    }
}
